File: sdk/go/lookup.go
package oati

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"sync"
	"time"
)

var RecordTypes = []string{"organisation", "agent", "passport", "mandate", "receipt", "issuer", "key", "revocation", "service", "profile"}

type OrganisationDiscovery struct {
	OrganisationID string             `json:"organisation_id"`
	Services       []DiscoveredRecord `json:"services"`
	Profiles       []DiscoveredRecord `json:"profiles"`
}
type DiscoveredRecord struct {
	Record   PublicRecord   `json:"record"`
	Document map[string]any `json:"document"`
}
type FederationDocument struct {
	OATIVersion   string   `json:"oati_version"`
	Organisations []string `json:"organisations"`
	Resolvers     []string `json:"resolvers"`
	ExpiresAt     string   `json:"expires_at,omitempty"`
}

type LookupError struct {
	Code       string
	Status     int
	RetryAfter time.Duration
	Err        error
}

func (e *LookupError) Error() string {
	if e.Err != nil {
		return e.Code + ": " + e.Err.Error()
	}
	return e.Code
}

type RateLimit struct {
	Limit, Remaining int
	ResetAt          string
	RetryAfter       time.Duration
}
type LookupResult struct {
	Record             PublicRecord
	ResolverURL, Cache string
	RateLimit          RateLimit
}
type LookupState struct {
	State  string
	Record *PublicRecord
	Err    error
}
type cacheEntry struct {
	record         *PublicRecord
	resolver, etag string
	rate           RateLimit
	expires        time.Time
	notFound       bool
}
type ResolverClient struct {
	ResolverURLs                     []string
	HTTPClient                       *http.Client
	MaxRetries                       int
	BaseDelay, CacheTTL, NegativeTTL time.Duration
	mu                               sync.Mutex
	cache                            map[string]cacheEntry
}

func NewResolverClient(urls ...string) *ResolverClient {
	if len(urls) == 0 {
		urls = []string{"https://api.intelliger.ai/oati/v1"}
	}
	return &ResolverClient{ResolverURLs: urls, HTTPClient: &http.Client{Timeout: 5 * time.Second}, MaxRetries: 2, BaseDelay: 200 * time.Millisecond, CacheTTL: time.Minute, NegativeTTL: 10 * time.Second, cache: map[string]cacheEntry{}}
}
func (c *ResolverClient) Lookup(ctx context.Context, kind, id string) (PublicRecord, error) {
	result, err := c.LookupDetailed(ctx, kind, id, false)
	return result.Record, err
}
func (c *ResolverClient) LookupDetailed(ctx context.Context, kind, id string, reload bool) (LookupResult, error) {
	return c.lookupSelectedDetailed(ctx, kind, "id", id, reload)
}
func (c *ResolverClient) LookupRevocationByTarget(ctx context.Context, target string) (PublicRecord, error) {
	result, err := c.LookupRevocationByTargetDetailed(ctx, target, false)
	return result.Record, err
}
func (c *ResolverClient) LookupRevocationByTargetDetailed(ctx context.Context, target string, reload bool) (LookupResult, error) {
	return c.lookupSelectedDetailed(ctx, "revocation", "target", target, reload)
}
func (c *ResolverClient) DiscoverOrganisation(ctx context.Context, organisationID string) (OrganisationDiscovery, error) {
	if !strings.HasPrefix(organisationID, "oati:org:") {
		return OrganisationDiscovery{}, &LookupError{Code: "LOOKUP_BAD_REQUEST", Err: fmt.Errorf("valid organisation id required")}
	}
	var last error
	for _, resolver := range c.ResolverURLs {
		endpoint := strings.TrimSuffix(resolver, "/") + "/discovery?organisation_id=" + url.QueryEscape(organisationID)
		request, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
		if err != nil {
			return OrganisationDiscovery{}, err
		}
		request.Header.Set("Accept", "application/json")
		response, err := c.http().Do(request)
		if err != nil {
			last = &LookupError{Code: "LOOKUP_UNAVAILABLE", Err: err}
			continue
		}
		var payload struct {
			OrganisationID string         `json:"organisation_id"`
			Services       []PublicRecord `json:"services"`
			Profiles       []PublicRecord `json:"profiles"`
		}
		decodeErr := json.NewDecoder(response.Body).Decode(&payload)
		response.Body.Close()
		if response.StatusCode == http.StatusNotFound {
			return OrganisationDiscovery{}, &LookupError{Code: "LOOKUP_NOT_FOUND", Status: 404}
		}
		if response.StatusCode < 200 || response.StatusCode >= 300 {
			last = &LookupError{Code: "LOOKUP_UNAVAILABLE", Status: response.StatusCode}
			continue
		}
		if decodeErr != nil || payload.OrganisationID != organisationID {
			return OrganisationDiscovery{}, &LookupError{Code: "LOOKUP_INVALID_RESPONSE", Err: decodeErr}
		}
		result := OrganisationDiscovery{OrganisationID: organisationID}
		for _, item := range payload.Services {
			decoded, decodeError := decodeDiscoveryRecord(item, "service", organisationID)
			if decodeError != nil {
				return OrganisationDiscovery{}, decodeError
			}
			result.Services = append(result.Services, decoded)
		}
		for _, item := range payload.Profiles {
			decoded, decodeError := decodeDiscoveryRecord(item, "profile", organisationID)
			if decodeError != nil {
				return OrganisationDiscovery{}, decodeError
			}
			result.Profiles = append(result.Profiles, decoded)
		}
		profiles := map[string]bool{}
		for _, item := range result.Profiles {
			profiles[item.Record.ID] = true
		}
		for _, item := range result.Services {
			for _, profile := range stringsFrom(item.Document["accepted_profiles"]) {
				if !profiles[profile] {
					return OrganisationDiscovery{}, &LookupError{Code: "LOOKUP_INVALID_RESPONSE", Err: fmt.Errorf("service references unpublished profile %s", profile)}
				}
			}
		}
		return result, nil
	}
	return OrganisationDiscovery{}, last
}
func (c *ResolverClient) DiscoverFederated(ctx context.Context, domain, organisationID string) (OrganisationDiscovery, error) {
	candidate := domain
	if !strings.Contains(candidate, "://") {
		candidate = "https://" + candidate
	}
	origin, err := url.Parse(candidate)
	if err != nil || origin.Scheme != "https" || origin.Hostname() == "" || origin.User != nil || (origin.Path != "" && origin.Path != "/") || origin.RawQuery != "" || origin.Fragment != "" {
		return OrganisationDiscovery{}, &LookupError{Code: "LOOKUP_BAD_REQUEST", Err: fmt.Errorf("federation requires a bare HTTPS domain")}
	}
	request, _ := http.NewRequestWithContext(ctx, http.MethodGet, origin.Scheme+"://"+origin.Host+"/.well-known/oati", nil)
	request.Header.Set("Accept", "application/json")
	response, err := c.http().Do(request)
	if err != nil {
		return OrganisationDiscovery{}, &LookupError{Code: "LOOKUP_UNAVAILABLE", Err: err}
	}
	defer response.Body.Close()
	if response.StatusCode == http.StatusNotFound {
		return OrganisationDiscovery{}, &LookupError{Code: "LOOKUP_NOT_FOUND", Status: 404}
	}
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return OrganisationDiscovery{}, &LookupError{Code: "LOOKUP_UNAVAILABLE", Status: response.StatusCode}
	}
	var document FederationDocument
	if json.NewDecoder(response.Body).Decode(&document) != nil || document.OATIVersion != "1.0" || !containsString(document.Organisations, organisationID) || len(document.Resolvers) == 0 || expired(document.ExpiresAt) {
		return OrganisationDiscovery{}, &LookupError{Code: "LOOKUP_INVALID_RESPONSE"}
	}
	for _, resolver := range document.Resolvers {
		parsed, parseErr := url.Parse(resolver)
		if parseErr != nil || parsed.Scheme != "https" || parsed.Hostname() == "" {
			return OrganisationDiscovery{}, &LookupError{Code: "LOOKUP_INVALID_RESPONSE"}
		}
	}
	federated := NewResolverClient(document.Resolvers...)
	federated.HTTPClient = c.HTTPClient
	federated.MaxRetries = c.MaxRetries
	federated.BaseDelay = c.BaseDelay
	federated.CacheTTL = c.CacheTTL
	federated.NegativeTTL = c.NegativeTTL
	return federated.DiscoverOrganisation(ctx, organisationID)
}
func (c *ResolverClient) lookupSelectedDetailed(ctx context.Context, kind, selector, value string, reload bool) (LookupResult, error) {
	if !containsString(RecordTypes, kind) || value == "" {
		return LookupResult{}, fmt.Errorf("valid record type and selector value required")
	}
	key := kind + "\x00" + selector + "\x00" + value
	c.mu.Lock()
	cached, ok := c.cache[key]
	c.mu.Unlock()
	if !reload && ok && cached.expires.After(time.Now()) {
		if cached.notFound {
			return LookupResult{}, &LookupError{Code: "LOOKUP_NOT_FOUND", Status: 404}
		}
		return LookupResult{Record: *cached.record, ResolverURL: cached.resolver, Cache: "hit", RateLimit: cached.rate}, nil
	}
	var last error
	for _, resolver := range c.ResolverURLs {
		reusable := cacheEntry{}
		if ok && cached.resolver == resolver {
			reusable = cached
		}
		result, etag, rate, revalidated, err := c.retry(ctx, resolver, kind, selector, value, reusable)
		if err == nil {
			c.store(key, cacheEntry{record: &result, resolver: resolver, etag: etag, rate: rate, expires: time.Now().Add(c.ttl())})
			cacheState := "miss"
			if revalidated {
				cacheState = "revalidated"
			}
			return LookupResult{Record: result, ResolverURL: resolver, Cache: cacheState, RateLimit: rate}, nil
		}
		last = err
		if lookup, yes := err.(*LookupError); yes && lookup.Code == "LOOKUP_NOT_FOUND" {
			c.store(key, cacheEntry{resolver: resolver, expires: time.Now().Add(c.negativeTTL()), notFound: true})
			return LookupResult{}, err
		}
	}
	return LookupResult{}, last
}
func (c *ResolverClient) LookupState(ctx context.Context, kind, id string) LookupState {
	record, err := c.Lookup(ctx, kind, id)
	if err != nil {
		state := "unavailable"
		if e, ok := err.(*LookupError); ok && e.Code == "LOOKUP_NOT_FOUND" {
			state = "not_found"
		}
		return LookupState{State: state, Err: err}
	}
	state := "found"
	if record.ProofStatus == "invalid" {
		state = "invalid_proof"
	} else if record.ProofStatus == "unknown" {
		state = "unknown"
	} else if record.ProofStatus == "unavailable" {
		state = "unavailable"
	}
	return LookupState{State: state, Record: &record}
}
func (c *ResolverClient) ClearCache() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.cache = map[string]cacheEntry{}
}
func (c *ResolverClient) retry(ctx context.Context, resolver, kind, selector, value string, cached cacheEntry) (PublicRecord, string, RateLimit, bool, error) {
	var last error
	for attempt := 0; attempt <= c.MaxRetries; attempt++ {
		if attempt > 0 {
			delay := c.BaseDelay * time.Duration(1<<(attempt-1))
			if lookup, ok := last.(*LookupError); ok && lookup.RetryAfter > 0 {
				delay = lookup.RetryAfter
			}
			select {
			case <-ctx.Done():
				return PublicRecord{}, "", RateLimit{}, false, contextLookupError(ctx)
			case <-time.After(delay):
			}
		}
		record, etag, rate, revalidated, err := c.request(ctx, resolver, kind, selector, value, cached)
		if err == nil {
			return record, etag, rate, revalidated, nil
		}
		last = err
		lookup, ok := err.(*LookupError)
		if !ok || (lookup.Code != "LOOKUP_UNAVAILABLE" && lookup.Code != "LOOKUP_RATE_LIMITED" && lookup.Code != "LOOKUP_TIMEOUT") {
			break
		}
	}
	return PublicRecord{}, "", RateLimit{}, false, last
}
func (c *ResolverClient) request(ctx context.Context, resolver, kind, selector, value string, cached cacheEntry) (PublicRecord, string, RateLimit, bool, error) {
	endpoint := strings.TrimSuffix(resolver, "/") + "/lookup?type=" + url.QueryEscape(kind) + "&" + selector + "=" + url.QueryEscape(value)
	request, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
	if err != nil {
		return PublicRecord{}, "", RateLimit{}, false, err
	}
	request.Header.Set("Accept", "application/json")
	if cached.etag != "" {
		request.Header.Set("If-None-Match", cached.etag)
	}
	response, err := c.http().Do(request)
	if err != nil {
		code := "LOOKUP_UNAVAILABLE"
		if ctx.Err() != nil {
			return PublicRecord{}, "", RateLimit{}, false, contextLookupError(ctx)
		}
		return PublicRecord{}, "", RateLimit{}, false, &LookupError{Code: code, Err: err}
	}
	defer response.Body.Close()
	rate := rateLimit(response)
	if response.StatusCode == http.StatusNotModified && cached.record != nil {
		return *cached.record, cached.etag, rate, true, nil
	}
	if response.StatusCode == http.StatusNotFound {
		return PublicRecord{}, "", rate, false, &LookupError{Code: "LOOKUP_NOT_FOUND", Status: 404}
	}
	if response.StatusCode == http.StatusTooManyRequests {
		return PublicRecord{}, "", rate, false, &LookupError{Code: "LOOKUP_RATE_LIMITED", Status: 429, RetryAfter: rate.RetryAfter}
	}
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return PublicRecord{}, "", rate, false, &LookupError{Code: "LOOKUP_UNAVAILABLE", Status: response.StatusCode}
	}
	var record PublicRecord
	if err := json.NewDecoder(response.Body).Decode(&record); err != nil {
		return PublicRecord{}, "", rate, false, &LookupError{Code: "LOOKUP_INVALID_RESPONSE", Err: err}
	}
	selectorMatches := record.ID == value
	if selector == "target" {
		status := record.PublicAttributes["revocation_status"]
		selectorMatches = record.Type == "revocation" && record.PublicAttributes["target"] == value && containsString([]string{"good", "suspended", "revoked", "unknown"}, status)
	}
	if record.Type != kind || !selectorMatches {
		return PublicRecord{}, "", rate, false, &LookupError{Code: "LOOKUP_INVALID_RESPONSE"}
	}
	if record.Status == "" || record.Issuer == "" || !containsString([]string{"verified", "invalid", "unavailable", "unknown"}, record.ProofStatus) || record.PublicAttributes == nil {
		return PublicRecord{}, "", rate, false, &LookupError{Code: "LOOKUP_INVALID_RESPONSE"}
	}
	if record.Type == "key" {
		if record.IssuedAt == "" {
			record.IssuedAt = record.PublicAttributes["valid_from"]
		}
		if record.ExpiresAt == "" {
			record.ExpiresAt = record.PublicAttributes["valid_until"]
		}
		if record.IssuedAt == "" || record.ExpiresAt == "" || record.PublicAttributes["controller"] == "" || !containsString([]string{"EdDSA", "ES256"}, record.PublicAttributes["algorithm"]) || record.PublicAttributes["public_key_jwk"] == "" {
			return PublicRecord{}, "", rate, false, &LookupError{Code: "LOOKUP_INVALID_RESPONSE"}
		}
	}
	if record.Type == "service" || record.Type == "profile" {
		if _, decodeErr := decodeDiscoveryRecord(record, record.Type, record.OrganisationID); decodeErr != nil {
			return PublicRecord{}, "", rate, false, decodeErr
		}
	}
	return record, response.Header.Get("ETag"), rate, false, nil
}
func contextLookupError(ctx context.Context) *LookupError {
	code := "LOOKUP_TIMEOUT"
	if errors.Is(ctx.Err(), context.Canceled) {
		code = "LOOKUP_CANCELLED"
	}
	return &LookupError{Code: code, Err: ctx.Err()}
}
func expired(value string) bool {
	if value == "" {
		return false
	}
	parsed, err := time.Parse(time.RFC3339, value)
	return err != nil || !parsed.After(time.Now())
}
func decodeDiscoveryRecord(record PublicRecord, kind, organisationID string) (DiscoveredRecord, error) {
	if record.Type != kind || record.OrganisationID != organisationID || record.Status != "active" || record.ProofStatus != "verified" {
		return DiscoveredRecord{}, &LookupError{Code: "LOOKUP_INVALID_RESPONSE", Err: fmt.Errorf("untrusted discovery record")}
	}
	raw := record.PublicAttributes["document"]
	var document map[string]any
	decoder := json.NewDecoder(strings.NewReader(raw))
	decoder.UseNumber()
	if raw == "" || decoder.Decode(&document) != nil {
		return DiscoveredRecord{}, &LookupError{Code: "LOOKUP_INVALID_RESPONSE", Err: fmt.Errorf("invalid discovery document")}
	}
	if str(document["id"]) != record.ID || str(document["organisation_id"]) != organisationID || str(document["issuer"]) != record.Issuer || str(document["status"]) != "active" || expired(record.ExpiresAt) || expired(str(document["expires_at"])) {
		return DiscoveredRecord{}, &LookupError{Code: "LOOKUP_INVALID_RESPONSE", Err: fmt.Errorf("invalid, mismatched, or expired discovery document")}
	}
	if kind == "service" {
		if str(document["oati_version"]) != "1.0" || len(list(document["endpoints"])) == 0 || document["accepted_profiles"] == nil {
			return DiscoveredRecord{}, &LookupError{Code: "LOOKUP_INVALID_RESPONSE"}
		}
	} else if str(document["oati_version"]) != "1.0" || str(document["schema_uri"]) == "" || str(document["digest"]) == "" {
		return DiscoveredRecord{}, &LookupError{Code: "LOOKUP_INVALID_RESPONSE"}
	}
	return DiscoveredRecord{Record: record, Document: document}, nil
}
func (c *ResolverClient) store(key string, value cacheEntry) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.cache == nil {
		c.cache = map[string]cacheEntry{}
	}
	c.cache[key] = value
}
func (c *ResolverClient) http() *http.Client {
	if c.HTTPClient == nil {
		return &http.Client{Timeout: 5 * time.Second}
	}
	return c.HTTPClient
}
func (c *ResolverClient) ttl() time.Duration {
	if c.CacheTTL == 0 {
		return time.Minute
	}
	return c.CacheTTL
}
func (c *ResolverClient) negativeTTL() time.Duration {
	if c.NegativeTTL == 0 {
		return 10 * time.Second
	}
	return c.NegativeTTL
}
func rateLimit(r *http.Response) RateLimit {
	limit, _ := strconv.Atoi(r.Header.Get("X-RateLimit-Limit"))
	remaining, _ := strconv.Atoi(r.Header.Get("X-RateLimit-Remaining"))
	retry, _ := strconv.Atoi(r.Header.Get("Retry-After"))
	return RateLimit{Limit: limit, Remaining: remaining, ResetAt: r.Header.Get("X-RateLimit-Reset"), RetryAfter: time.Duration(retry) * time.Second}
}

File: sdk/python/src/oati/lookup.py
from __future__ import annotations
import json,time
from datetime import datetime,timezone
from dataclasses import dataclass
from typing import Callable,Any
from urllib.error import HTTPError,URLError
from urllib.parse import urlencode,urlparse
from urllib.request import Request,urlopen
from .core import validate_schema

RECORD_TYPES=("organisation","agent","passport","mandate","receipt","issuer","key","revocation","service","profile")
class LookupError(RuntimeError):
    def __init__(self,code:str,message:str,*,status:int|None=None,retry_after:float|None=None):super().__init__(message);self.code=code;self.status=status;self.retry_after=retry_after
@dataclass(frozen=True)
class LookupResult:
    record:dict[str,Any];resolver_url:str;cache:str;rate_limit:dict[str,Any]
@dataclass
class _CacheEntry:
    record:dict[str,Any]|None;resolver_url:str;expires_at:float;etag:str|None=None;not_found:bool=False

class LookupClient:
    def __init__(self,resolver_urls:list[str]|None=None,*,timeout:float=5,max_retries:int=2,base_delay:float=.2,ttl:float=60,negative_ttl:float=10,opener:Callable[...,Any]=urlopen):
        self.resolver_urls=resolver_urls or ["https://api.intelliger.ai/oati/v1"];self.timeout=timeout;self.max_retries=max_retries;self.base_delay=base_delay;self.ttl=ttl;self.negative_ttl=negative_ttl;self.opener=opener;self._cache:dict[str,_CacheEntry]={}
    def lookup(self,record_type:str,record_id:str,*,reload:bool=False,no_store:bool=False)->dict[str,Any]:return self.lookup_detailed(record_type,record_id,reload=reload,no_store=no_store).record
    def lookup_detailed(self,record_type:str,record_id:str,*,reload:bool=False,no_store:bool=False)->LookupResult:
        return self._lookup_selected(record_type,"id",record_id,reload=reload,no_store=no_store)
    def lookup_revocation_by_target(self,target:str,*,reload:bool=False,no_store:bool=False)->dict[str,Any]:return self.lookup_revocation_by_target_detailed(target,reload=reload,no_store=no_store).record
    def lookup_revocation_by_target_detailed(self,target:str,*,reload:bool=False,no_store:bool=False)->LookupResult:return self._lookup_selected("revocation","target",target,reload=reload,no_store=no_store)
    def discover_organisation(self,organisation_id:str)->dict[str,Any]:
        if not organisation_id.startswith("oati:org:"):raise LookupError("LOOKUP_BAD_REQUEST","valid organisation id required")
        last=None
        for resolver in self.resolver_urls:
            endpoint=resolver.rstrip("/")+"/discovery?"+urlencode({"organisation_id":organisation_id})
            try:
                with self.opener(Request(endpoint,headers={"Accept":"application/json"}),timeout=self.timeout) as response:value=json.load(response)
            except HTTPError as error:
                if error.code==404:raise LookupError("LOOKUP_NOT_FOUND","discovery not found",status=404) from error
                last=LookupError("LOOKUP_UNAVAILABLE",f"discovery returned HTTP {error.code}",status=error.code);continue
            except (URLError,OSError) as error:last=LookupError("LOOKUP_UNAVAILABLE",str(error));continue
            if not isinstance(value,dict) or value.get("organisation_id")!=organisation_id or not isinstance(value.get("services"),list) or not isinstance(value.get("profiles"),list):raise LookupError("LOOKUP_INVALID_RESPONSE","invalid discovery response")
            services=[_decode_discovery(record,"service",organisation_id) for record in value["services"]]
            profiles=[_decode_discovery(record,"profile",organisation_id) for record in value["profiles"]]
            profile_ids={item["record"]["id"] for item in profiles}
            if any(profile not in profile_ids for item in services for profile in item["document"]["accepted_profiles"]):raise LookupError("LOOKUP_INVALID_RESPONSE","service references an unpublished profile")
            return {"organisation_id":organisation_id,"services":services,"profiles":profiles}
        raise last or LookupError("LOOKUP_UNAVAILABLE","no resolver was available")
    def discover_federated(self,domain:str,organisation_id:str)->dict[str,Any]:
        parsed=urlparse(domain if "://" in domain else "https://"+domain)
        if parsed.scheme!="https" or not parsed.hostname or parsed.username or parsed.password or parsed.path not in ("", "/") or parsed.query or parsed.fragment:raise LookupError("LOOKUP_BAD_REQUEST","federation requires a bare HTTPS domain")
        endpoint=f"https://{parsed.netloc}/.well-known/oati"
        try:
            with self.opener(Request(endpoint,headers={"Accept":"application/json"}),timeout=self.timeout) as response:document=json.load(response)
        except HTTPError as error:raise LookupError("LOOKUP_NOT_FOUND" if error.code==404 else "LOOKUP_UNAVAILABLE","federation lookup failed",status=error.code) from error
        except (URLError,OSError,TimeoutError) as error:raise LookupError("LOOKUP_UNAVAILABLE",str(error)) from error
        valid=isinstance(document,dict) and document.get("oati_version")=="1.0" and organisation_id in document.get("organisations",[]) and isinstance(document.get("resolvers"),list) and bool(document["resolvers"])
        if not valid or any(urlparse(item).scheme!="https" for item in document.get("resolvers",[])) or _expired(document.get("expires_at")):raise LookupError("LOOKUP_INVALID_RESPONSE","invalid, unrelated, or expired federation document")
        return LookupClient(document["resolvers"],timeout=self.timeout,max_retries=self.max_retries,base_delay=self.base_delay,ttl=self.ttl,negative_ttl=self.negative_ttl,opener=self.opener).discover_organisation(organisation_id)
    def _lookup_selected(self,record_type:str,selector:str,value:str,*,reload:bool,no_store:bool)->LookupResult:
        if record_type not in RECORD_TYPES or not value:raise ValueError("valid record type and selector value are required")
        key=record_type+"\0"+selector+"\0"+value;cached=self._cache.get(key)
        if not reload and not no_store and cached and cached.expires_at>time.monotonic():
            if cached.not_found:raise LookupError("LOOKUP_NOT_FOUND","OATI record was not found",status=404)
            return LookupResult(dict(cached.record or {}),cached.resolver_url,"hit",{})
        last:LookupError|None=None
        for resolver in self.resolver_urls:
            try:
                reusable=cached if cached and cached.resolver_url==resolver else None
                result,etag,rate,revalidated=self._request_with_retry(resolver,record_type,selector,value,reusable)
                if not no_store:self._cache[key]=_CacheEntry(result,resolver,time.monotonic()+self.ttl,etag)
                return LookupResult(result,resolver,"revalidated" if revalidated else "miss",rate)
            except LookupError as error:
                last=error
                if error.code=="LOOKUP_NOT_FOUND":
                    if not no_store:self._cache[key]=_CacheEntry(None,resolver,time.monotonic()+self.negative_ttl,not_found=True)
                    raise
        raise last or LookupError("LOOKUP_UNAVAILABLE","no resolver was available")
    def lookup_state(self,record_type:str,record_id:str)->dict[str,Any]:
        try:
            result=self.lookup_detailed(record_type,record_id);proof=result.record.get("proof_status")
            return {"state":"invalid_proof" if proof=="invalid" else "unavailable" if proof=="unavailable" else "unknown" if proof=="unknown" else "found","record":result.record,"response":result}
        except LookupError as error:return {"state":"not_found" if error.code=="LOOKUP_NOT_FOUND" else "unavailable","error":error}
    def clear_cache(self):self._cache.clear()
    def _request_with_retry(self,resolver,record_type,selector,value,cached):
        last=None
        for attempt in range(self.max_retries+1):
            if attempt:time.sleep(min(self.base_delay*2**(attempt-1),5) if not last or last.retry_after is None else min(last.retry_after,5))
            try:return self._request(resolver,record_type,selector,value,cached)
            except LookupError as error:
                last=error
                if error.code not in ("LOOKUP_RATE_LIMITED","LOOKUP_UNAVAILABLE","LOOKUP_TIMEOUT") or attempt==self.max_retries:raise
        raise last
    def _request(self,resolver,record_type,selector,selector_value,cached):
        endpoint=resolver.rstrip("/")+"/lookup?"+urlencode({"type":record_type,selector:selector_value});headers={"Accept":"application/json"}
        if cached and cached.etag:headers["If-None-Match"]=cached.etag
        try:
            with self.opener(Request(endpoint,headers=headers),timeout=self.timeout) as response:
                if getattr(response,"status",200)==304 and cached and cached.record:return dict(cached.record),cached.etag,{},True
                value=json.load(response);response_headers=response.headers
        except HTTPError as error:
            retry=_retry_after(error.headers.get("Retry-After")) if error.headers else None
            code="LOOKUP_NOT_FOUND" if error.code==404 else "LOOKUP_RATE_LIMITED" if error.code==429 else "LOOKUP_UNAVAILABLE"
            raise LookupError(code,f"lookup returned HTTP {error.code}",status=error.code,retry_after=retry) from error
        except TimeoutError as error:raise LookupError("LOOKUP_TIMEOUT",str(error)) from error
        except (URLError,OSError) as error:raise LookupError("LOOKUP_UNAVAILABLE",str(error)) from error
        matches=isinstance(value,dict) and (value.get("id")==selector_value if selector=="id" else value.get("type")=="revocation" and value.get("public_attributes",{}).get("target")==selector_value and value.get("public_attributes",{}).get("revocation_status") in ("good","suspended","revoked","unknown"))
        if not matches or value.get("type")!=record_type:raise LookupError("LOOKUP_INVALID_RESPONSE","resolver returned a mismatched record")
        value=_normalize_record(value)
        rate={"limit":_integer(response_headers.get("X-RateLimit-Limit")),"remaining":_integer(response_headers.get("X-RateLimit-Remaining")),"reset_at":response_headers.get("X-RateLimit-Reset")}
        if record_type=="key":
            attributes=value.get("public_attributes",{})
            value={**value,"issued_at":value.get("issued_at") or attributes.get("valid_from"),"expires_at":value.get("expires_at") or attributes.get("valid_until")}
        return value,response_headers.get("ETag"),{key:value for key,value in rate.items() if value is not None},False

def lookup(base_url:str,record_type:str,record_id:str,timeout:float=5.0)->dict:return LookupClient([base_url],timeout=timeout,max_retries=0).lookup(record_type,record_id)
def _integer(value):
    try:return int(value)
    except (TypeError,ValueError):return None
def _retry_after(value):
    try:return float(value)
    except (TypeError,ValueError):return None
def _expired(value):
    if not value:return False
    try:return datetime.fromisoformat(value.replace("Z","+00:00"))<=datetime.now(timezone.utc)
    except (TypeError,ValueError):return True
def _decode_discovery(record,kind,organisation_id):
    if not isinstance(record,dict) or record.get("type")!=kind or record.get("organisation_id")!=organisation_id or record.get("status")!="active" or record.get("proof_status")!="verified":raise LookupError("LOOKUP_INVALID_RESPONSE","untrusted discovery record")
    try:document=json.loads(record.get("public_attributes",{}).get("document",""))
    except (TypeError,ValueError) as error:raise LookupError("LOOKUP_INVALID_RESPONSE","invalid discovery document") from error
    if not isinstance(document,dict) or document.get("id")!=record.get("id") or document.get("organisation_id")!=organisation_id or document.get("issuer")!=record.get("issuer") or document.get("status")!="active":raise LookupError("LOOKUP_INVALID_RESPONSE","discovery document does not match record")
    if validate_schema("serviceDiscovery" if kind=="service" else "profileDiscovery",document) or _expired(record.get("expires_at")) or _expired(document.get("expires_at")):raise LookupError("LOOKUP_INVALID_RESPONSE","invalid or expired discovery document")
    return {"record":record,"document":document}
def _normalize_record(value):
    attributes=value.get("public_attributes")
    if not isinstance(attributes,dict) or not isinstance(value.get("status"),str) or not isinstance(value.get("issuer"),str) or value.get("proof_status") not in ("verified","invalid","unavailable","unknown"):raise LookupError("LOOKUP_INVALID_RESPONSE","resolver returned an invalid public record")
    if not all(isinstance(item,str) for item in attributes.values()):raise LookupError("LOOKUP_INVALID_RESPONSE","public attributes must be strings")
    if value.get("type")=="key":
        value={**value,"issued_at":value.get("issued_at") or attributes.get("valid_from"),"expires_at":value.get("expires_at") or attributes.get("valid_until")}
        if not value.get("issued_at") or not value.get("expires_at") or not attributes.get("controller") or attributes.get("algorithm") not in ("EdDSA","ES256") or not attributes.get("public_key_jwk"):raise LookupError("LOOKUP_INVALID_RESPONSE","resolver returned an invalid key record")
    if value.get("type") in ("service","profile"):_decode_discovery(value,value["type"],value.get("organisation_id"))
    return value
